File: Spooky2Net/src/Spooky2.Core/Models/ScanParameters.cs
namespace Spooky2.Core.Models;

public sealed record ScanParameters
{
    public double StartFrequency { get; init; } = 41000;
    public double EndFrequency { get; init; } = 1800000;
    public bool UsePercentageStep { get; init; } = true;
    public double StepSizeHz { get; init; } = 100;
    public double StepSizePercent { get; init; } = 0.025;
    public int MaxHits { get; init; } = 10;
    /// <summary>Primary RA window size (BFB_RA_Window_1).</summary>
    public int RaWindow { get; init; } = 20;
    /// <summary>Secondary RA window size (BFB_RA_Window_2). 0 = same as RaWindow.</summary>
    public int RaWindow2 { get; init; }
    /// <summary>Use the secondary (retentive) RA window for detection.</summary>
    public bool UseRetentiveWindow { get; init; }
    /// <summary>Use peak detection instead of running average.</summary>
    public bool CalculateUsingPeak { get; init; }
    public int SamplesPerStep { get; init; } = 1;
    public int StartDelayMs { get; init; } = 200;
    public double MinReadDelaySeconds { get; init; } = 0.07;
    public bool DetectMax { get; init; } = true;
    public bool DetectMin { get; init; }
    /// <summary>Use current (mA) sensor for hit detection.
    /// Despite the GX Hunt and Kill preset saying BFB_Detect_mA=True,
    /// empirical testing against real scan data proves the original uses
    /// angle/phase for detection — angle finds all 10 expected hits across
    /// all parameter combinations, while current produces spurious hits at
    /// 980K/1150K Hz that mask weaker real hits.</summary>
    public bool UseCurrent { get; init; }
    public bool UseAngle { get; init; } = true;
    public int Loops { get; init; } = 1;
    public double Threshold { get; init; }
    public bool ContinueRefining { get; init; } = true;
    public bool RunHitsAfterScan { get; init; } = true;
    public int RunOnGeneratorId { get; init; }
    public double DwellSeconds { get; init; } = 180;
    public string LogName { get; init; } = "";

    // Amplitude ramp-up (from preset: Enable_Amplitude_RampUp, Ramp_Amplitude_Up_Rate)
    /// <summary>Enable gradual amplitude ramp from near-zero to target before scanning.</summary>
    public bool EnableAmplitudeRampUp { get; init; } = true;
    /// <summary>Number of ramp steps (verified from dump: 330 steps, formula round((i+1)*target/330)).</summary>
    public int RampSteps { get; init; } = 330;
    /// <summary>Target amplitude in centivolt (2000 = 20.00V).</summary>
    public int TargetAmplitudeCv { get; init; } = 2000;
    /// <summary>Enable amplitude ramp-down after scan.</summary>
    public bool EnableAmplitudeRampDown { get; init; } = true;
    /// <summary>Number of baseline :r11/:r12 read pairs before sweeping.
    /// Dump shows 203 pairs (+ 1 initial standalone :r11 = 407 total reads).</summary>
    public int BaselineReadCount { get; init; } = 203;
    /// <summary>Minimum frequency separation between hits as a percentage (0.025 = one step).
    /// When a new hit is within this distance of an existing hit, the one with higher deviation
    /// is kept and the other is discarded. Prevents cluster monopolization of hit slots.
    /// Verified from original VB6 output: nearest distinct hits are ~0.05% apart.</summary>
    public double MinHitSeparationPercent { get; init; } = 0.05;
}

File: spooky2-huntkill-android/core/src/main/kotlin/com/spooky2/huntkill/core/model/ScanParameters.kt
package com.spooky2.huntkill.core.model

/**
 * Hunt & Kill scan configuration.
 *
 * Verbatim port of the C# reference `Spooky2.Core.Models.ScanParameters`.
 * Default values are copied exactly from the C# record and reflect the
 * GX Hunt and Kill preset cross-checked against the recorded scan dumps.
 */
data class ScanParameters(
    val startFrequency: Double = 41000.0,
    val endFrequency: Double = 1800000.0,
    val usePercentageStep: Boolean = true,
    val stepSizeHz: Double = 100.0,
    val stepSizePercent: Double = 0.025,
    val maxHits: Int = 10,
    /** Primary RA window size (BFB_RA_Window_1). */
    val raWindow: Int = 20,
    /** Secondary RA window size (BFB_RA_Window_2). 0 = same as [raWindow]. */
    val raWindow2: Int = 0,
    /** Use the secondary (retentive) RA window for detection. */
    val useRetentiveWindow: Boolean = false,
    /** Use peak detection instead of running average. */
    val calculateUsingPeak: Boolean = false,
    val samplesPerStep: Int = 1,
    val startDelayMs: Int = 200,
    /**
     * Minimum sweep-step PERIOD (write-to-write), in seconds. Default 0.07 = 70 ms.
     *
     * Derived from the original software's timestamped serial dump, which sweeps
     * at 14-15 steps/s (≈70 ms/step). This is the minimum time between successive
     * `:w24` frequency writes — NOT an additive sleep. The per-step serial I/O
     * (frequency write + sensor reads) counts toward this period; the engine only
     * sleeps the remainder. 0 = unpaced fast path (test/replay).
     */
    val minReadDelaySeconds: Double = 0.07,
    val detectMax: Boolean = true,
    val detectMin: Boolean = false,
    /**
     * Detection settle warm-up tolerance (fraction of the window mean). Detection does
     * not score any sweep step until the SMA window has "settled": the window's
     * (max − min) ≤ [settleToleranceFraction] × windowMean. The first step (≥ [raWindow],
     * so the window is full) that satisfies this is `warmupStart`; scoring begins there.
     *
     * This is a one-time LEADING warm-up: it suppresses a generator/sensor startup
     * transient (the first few sweep readings sitting at the baseline level before they
     * JUMP to the settled level) from being selected as a false top hit. It does NOT
     * re-gate later steps — a real peak legitimately widens the window range. If no
     * settled window is found within a cap (5 × [raWindow]) the warm-up falls back to
     * [raWindow] so a genuinely noisy scan is never fully discarded. Default 0.01 (1%).
     */
    val settleToleranceFraction: Double = 0.01,
    /**
     * Use current (mA) sensor for hit detection.
     * Despite the GX Hunt and Kill preset saying BFB_Detect_mA=True,
     * empirical testing against real scan data proves the original uses
     * angle/phase for detection — angle finds all 10 expected hits across
     * all parameter combinations, while current produces spurious hits at
     * 980K/1150K Hz that mask weaker real hits.
     */
    val useCurrent: Boolean = false,
    val useAngle: Boolean = true,
    val loops: Int = 1,
    val threshold: Double = 0.0,
    val continueRefining: Boolean = true,
    /**
     * Generator ID for kill phase output. 0 = same generator as scan.
     * Maps to preset BFB_After_Scan_Run_On_Gen=0.
     */
    val runOnGeneratorId: Int = 0,
    /**
     * Kill phase dwell time per frequency. Default 180s (3 minutes) from dump analysis.
     * No preset field — derived from observed behavior.
     */
    val dwellSeconds: Double = 180.0,
    val logName: String = "",

    // Amplitude ramp-up (from preset: Enable_Amplitude_RampUp, Ramp_Amplitude_Up_Rate)
    /** Enable gradual amplitude ramp from near-zero to target before scanning. */
    val enableAmplitudeRampUp: Boolean = true,
    /**
     * Number of amplitude ramp steps. Default 330 from dump analysis.
     * Preset has Ramp_Amplitude_Up_Rate=4 but the mapping formula is unknown.
     * Verified from dump: 330 steps, formula round((i+1)*target/330).
     */
    val rampSteps: Int = 330,
    /**
     * Target amplitude in centivolt. Default 2000 = 20V.
     * Maps to preset Out1_Amplitude=20.
     */
    val targetAmplitudeCv: Int = 2000,
    /** Enable amplitude ramp-down after scan. */
    val enableAmplitudeRampDown: Boolean = true,
    /**
     * Number of baseline :r11/:r12 read pairs before sweeping.
     * Dump shows 203 pairs (+ 1 initial standalone :r11 = 407 total reads).
     */
    val baselineReadCount: Int = 203,

    // ── Dropout detection (cable disconnect / half-broken connection) ──
    /**
     * Minimum contiguous run length (steps) for the deviation heuristic to flag
     * a dropout. Short excursions (1-2 steps) are normal biofeedback variation;
     * a sustained plateau of >= this many steps is cable garbage.
     */
    val dropoutMinRunLength: Int = 3,
    /**
     * Fractional deviation from the surrounding rolling median above which a step
     * is considered an outlier candidate (0.10 = 10%). Tuned so the observed
     * failure (collapse to a far-away plateau) is caught but normal variation
     * (a few percent) is not.
     */
    val dropoutDeviationFraction: Double = 0.10,
    /**
     * Window radius (steps on each side) for the rolling median used by the
     * deviation heuristic. The median is taken over valid readings in this window.
     */
    val dropoutMedianWindow: Int = 25,
    /**
     * Consecutive failed reads after which the live "connection unstable" status
     * is surfaced (the sweep keeps going, values flagged).
     */
    val dropoutUnstableReadThreshold: Int = 5,
) {
    init {
        require(samplesPerStep > 0) { "samplesPerStep must be > 0, got $samplesPerStep" }
    }
}
